src/libraries/System.Linq/src/System/Linq/OrderBy.cs

```diff
@@ -140,14 +140,37 @@ public static partial class Enumerable
 
         /// <summary>Gets whether the results of an unstable sort will be observably the same as a stable sort.</summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static bool TypeIsImplicitlyStable<T>() =>
-            typeof(T) == typeof(sbyte) || typeof(T) == typeof(byte) ||
-            typeof(T) == typeof(int) || typeof(T) == typeof(uint) ||
-            typeof(T) == typeof(short) || typeof(T) == typeof(ushort) ||
-            typeof(T) == typeof(long) || typeof(T) == typeof(ulong) ||
-            typeof(T) == typeof(Int128) || typeof(T) == typeof(UInt128) ||
-            typeof(T) == typeof(nint) || typeof(T) == typeof(nuint) ||
-            typeof(T) == typeof(bool) || typeof(T) == typeof(char);
+        internal static bool TypeIsImplicitlyStable<T>()
+        {
+            // Check for integral primitive types that compare equally iff they have the same bit pattern.
+            // (bool is included because, even though technically it can have 256 different values, anything
+            // other than 0/1 is only producible using unsafe code.)
+            if (typeof(T) == typeof(sbyte) || typeof(T) == typeof(byte) || typeof(T) == typeof(bool) ||
+                typeof(T) == typeof(short) || typeof(T) == typeof(ushort) || typeof(T) == typeof(char) ||
+                typeof(T) == typeof(int) || typeof(T) == typeof(uint) ||
+                typeof(T) == typeof(long) || typeof(T) == typeof(ulong) ||
+                typeof(T) == typeof(Int128) || typeof(T) == typeof(UInt128) ||
+                typeof(T) == typeof(nint) || typeof(T) == typeof(nuint))
+            {
+                return true;
+            }
+
+            if (typeof(T).IsEnum)
+            {
+                // Check for all integral underlying types.
```

Comment:
it wouldn't work to move this check first, ie put T into a variable, and if it's an enum set that to the underlying type, and then writing all the type checks out once?
but I'm guessing that would take some codegen-time work to runtime work because directly comparing to typeof(T) is codegen time work and free at runtime?

Reply:
> but I'm guessing that would take some codegen-time work to runtime work because directly comparing to typeof(T) is codegen time work and free at runtime?

Correct, e.g.
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4axAMxNSDIQwDeNGgzNMdwCBEkMAkrjvjJy3hgD6ABTYAKAJQMALwAfAzeAGoyuLwQXAA8ACLYAJ4A8gBmAOowMADWwf4a1OYMpubaDFY29o7Orh4i+AI8/kGhjc0YkVDRsYkpGdl5BX5FJWVmANoAsjAYABYQACZOYpI+swvLq5KpYnyxuCwAggDmp7C40QowdlwuXLxcp34AuhMMTxgyXNi2zEhKtZbBEojF4gAVEYfYzFErmXjpBg+DDJMQwCDpHwQgKBQIMVHozE+XDAZLfAKEQgEtEYrE4oL4wl0nxkikMKk0on03FM2nEqqSSmED7w8zM4kMvFclm4RZQDDCmWS3nKrEcOXQRUc6kSnmMtU+MDzbBQYWisWGqV87k+L5KvXY1WOjj2nUWsWO62GySxF46q3O/nq33Pc1wy3i4NOg2Ou4YcikAAcDuj3sdAFV44mU+6I5HA7Ho48eKnbeni67S34PUZayViAB2AlQDgwMaWgC+JnzJURyK9fhYDgAolwOPga73zLCC+YIbSGNwljJJMknqcgoGWABxOZjicZrgrqBrjcL9GFevlZvL1fr55F22k8kwJV308PzfSx1st95ucCw/M9Hx/aNNQVd9j3vDcnxZDV5W1TlgK/ODiWNU1w0AyMUNgsDbTdZDoM/PCbXgwiRWnbCl2IkDvzI4lQ39IiTzotCQz9LDqPhXDQIYrESyQ6lePow1x3tDsxW7ahaybBh0j+XB2w+aTxnzGY5kWFZnA2TTtmcPYDi4I4zguVRrlue4ng3N4Pi+H4/iYchAUFfRRE6bpekhaF81nS0LxgGUt0HST4X7FFaWHXAD0nWs/LnZlgrTIc9wwGKj1Yr8AqvKiGFUy1ZMba8zES/DZT/MtArKgVX0q9jWWBLiEsXaqsQgoSgtanwEK1Oquows0AMA0r+LtUsAxGw0qyE4rOtGpi+tGjgFqG5r0Xq7Nk0Ww0sx4HMmoLSbHUE7bjum0YVJoTsgA=)

Reply:
> SharpLab

FYI it seems to work [after a small change.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4axAMxNSDIQwDeNGgzNMdwCBEkMAkrjvjJy3hgD6ABTYAKAJQMALwAfAzeAGoyuLwQXAA8ACLYAJ4A8gBmAOowMADWwf4a1OYMpubaDFY29o7Orh4i+AI8/kGhjc0YkVDRsYkpGdl5BX5FJWVmANoAsjAYABYQACZOYpI+swvLq5KpYnyxuCwAggDmp7C40QowdlwuXLxcp34AuhMMTxgyXNi2zEhKtZbBEojF4gAVEYfYzFErmXjpBg+DDJMQwCDpHwQgKBQIMVHozE+XDAZLfAKEQgEtEYrE4oL4wl0nxkikMKk0on03FM2nEqqSSmED7w8zM4kMvFclm4RZQDDCmWS3nKrEcOXQRUc6kSnmMtU+MDzbBQYWisWGqV87k+L5KvXY1WOjj2nUWsWO62GySxF46q3O/nq33Pc1wy3i4NOg2Ou4YcikAAcDuj3sdAFV44mU+6I5HA7Ho48eKnbeni67S34PUZayViAB2AlQDgwMaWgC+JnzJURyK9fhYDgAolwOPga73zLCC+YIbSGNwljJJMknqcgoGWABxOZjicZrgrqBrjcL9GFevlZvL1fr55F22k8kwJV308PzfSx1st95ucCw/M9Hx/aNNQVd9j3vDcnxZDV5W1TlgK/ODiWNU1w0AyMUNgsDbTdZDoM/PCbXgwiRWnbCl2IkDvzI4lQ39IiTzotCQz9LDqPhXDQIYrESyQ6lePow1x3tDsxW7ahaybBh0j+XB2w+aTxnzGY5kWFZnA2TTtmcPYDi4I4zguVRrlue4ng3N4Pi+H4/iYchAUFfRRE6bpekhaF81nS0LxgGUt0HST4X7FE0yHUdx0nWs/LnZlgsi3d9xio9WK/AKryohhVMtWTG2vMxEvw2U/zLQLSoFV8KvY1lgS4hLFyqrEIKEoKWp8BCtVqzqMLNADAJK/i7VLANhsNKshKKjqRqY3qRo4ebBqa9E6uzZMFsNLMeBzRqCwmx1BK2o6ptGFSaE7IA===)

Reply:
Nice. I'll simplify then.

Reply:
Nice! It's beautiful when that happens